core/editor/src/tools/ellipse.rs

```diff
@@ -1,12 +1,54 @@
-use crate::events::Event;
-use crate::tools::Tool;
+use crate::events::MouseKeys;
+use crate::events::{Event, Response};
+use crate::tools::{Fsm, Tool};
+use crate::Document;
 use document_core::Operation;
 
 #[derive(Default)]
-pub struct Ellipse;
+pub struct Ellipse {
+	state: EllipseToolState,
+}
 
 impl Tool for Ellipse {
-	fn handle_input(&mut self, event: Event) -> Vec<Operation> {
-		todo!();
+	fn handle_input(&mut self, event: &Event, document: &Document) -> (Vec<Response>, Vec<Operation>) {
+		let mut responses = Vec::new();
+		let mut operations = Vec::new();
+		self.state = self.state.transition(event, document, &mut responses, &mut operations);
+
+		(responses, operations)
+	}
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum EllipseToolState {
+	Ready,
+	LmbDown,
+	TransformSelected,
+}
+
+impl Default for EllipseToolState {
```

Comment:
Can we make this happen automatically for every tool by deriving a shared trait? `Ready` should always be the default trait, preferably without having to specify it on each tool. Or maybe we can pass a function or macro with a list of states and it adds the "Ready" state automatically.

Reply:
That's not as easy as it sounds since the `Ready` states are variants of different enums and the only thing they share is their name

Reply:
Ah, fair point. Perhaps we can optimize the cleanliness of this in the future but this should be a reasonable approach for now.